File: src/util/output_utils.c
/*
 * Output Utilities
 * Helper functions for formatting and emitting tool output
 */

#include "output_utils.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <bsd/string.h>

#ifndef TEST_BUILD
#include "../logger.h"
#include "../message_queue.h"
#else
// Test build stubs
#define LOG_ERROR(fmt, ...) ((void)0)
#define LOG_WARN(fmt, ...) ((void)0)
#endif

#include "../fallback_colors.h"

// Thread-local state for tool output
static _Thread_local TUIMessageQueue *g_active_tool_queue = NULL;
static _Thread_local int g_oneshot_mode = 0;

/**
 * Set the active tool queue for output
 */
void output_set_tool_queue(TUIMessageQueue *queue) {
    g_active_tool_queue = queue;
}

/**
 * Get the current tool queue
 */
TUIMessageQueue* output_get_tool_queue(void) {
    return g_active_tool_queue;
}

/**
 * Set oneshot mode flag
 */
void output_set_oneshot_mode(int enabled) {
    g_oneshot_mode = enabled;
}

/**
 * Get oneshot mode flag
 */
int output_get_oneshot_mode(void) {
    return g_oneshot_mode;
}

/**
 * Emit a line of tool output
 */
void tool_emit_line(const char *prefix, const char *text) {
    const char *safe_prefix = prefix ? prefix : "";
    const char *safe_text = text ? text : "";

    if (g_active_tool_queue) {
        size_t prefix_len = safe_prefix[0] ? strlen(safe_prefix) : 0;
        size_t text_len = strlen(safe_text);
        size_t extra = (prefix_len > 0 && text_len > 0) ? 1 : 0;
        size_t total = prefix_len + extra + text_len + 1;

        char *formatted = malloc(total);
        if (!formatted) {
            LOG_ERROR("Failed to allocate tool output buffer");
            return;
        }

        if (prefix_len > 0 && text_len > 0) {
            snprintf(formatted, total, "%s %s", safe_prefix, safe_text);
        } else if (prefix_len > 0) {
            snprintf(formatted, total, "%s", safe_prefix);
        } else {
            snprintf(formatted, total, "%s", safe_text);
        }

        if (post_tui_message(g_active_tool_queue, TUI_MSG_ADD_LINE, formatted) != 0) {
            LOG_WARN("Failed to post tool output to TUI queue");
        }
        free(formatted);
        return;
    }

    // In oneshot/subagent mode, suppress individual line output
    // Tool output will be captured and wrapped in HTML-style tags by the caller
    if (g_oneshot_mode) {
        return;
    }

    if (safe_prefix[0] && safe_text[0]) {
        printf("%s %s\n", safe_prefix, safe_text);
    } else if (safe_prefix[0]) {
        printf("%s\n", safe_prefix);
    } else {
        printf("%s\n", safe_text);
    }
    fflush(stdout);
}

/**
 * Emit a diff line with appropriate coloring
 */
void emit_diff_line(const char *line,
                    const char *add_color,
                    const char *remove_color) {
    if (!line) {
        return;
    }

    // Trim trailing newlines
    size_t len = strlen(line);
    while (len > 0 && (line[len - 1] == '\n' || line[len - 1] == '\r')) {
        len--;
    }

    if (len == 0) {
        return;
    }

    char *trimmed = strndup(line, len);
    if (!trimmed) {
        LOG_ERROR("Failed to allocate trimmed diff line");
        return;
    }

    // Print with indentation and color if applicable
    if (g_active_tool_queue) {
        // For TUI mode: just pass the line as-is
        // The TUI will detect diff prefixes and color appropriately
        tool_emit_line("", trimmed);
    } else if (!g_oneshot_mode) {
        // For non-TUI mode (direct stdout): use ANSI color codes
        // Skip output in oneshot/subagent mode as it will be captured in JSON
        const char *color = NULL;
        if (trimmed[0] == '+' && trimmed[1] != '+') {
            color = add_color;
        } else if (trimmed[0] == '-' && trimmed[1] != '-') {
            color = remove_color;
        }

        if (color) {
            printf("  %s%s%s\n", color, trimmed, ANSI_RESET);
        } else {
            printf("  %s\n", trimmed);
        }
    }

    free(trimmed);
}

File: zig/util/output_utils.zig
//! Output Utilities
//!
//! Idiomatic Zig replacements for src/util/output_utils.c
//!
//! The C version was tightly coupled to the TUI message queue, global thread-local
//! state, and the oneshot mode flag.  The full TUI integration belongs in Phase 9.
//!
//! This module provides the *pure formatting* primitives that are independent
//! of rendering destination.  A higher-level output coordinator (Phase 8/9)
//! will wire these to the TUI queue or direct stdout as appropriate.
//!
//! Key C→Zig translations:
//!   - `tool_emit_line`   → `formatToolLine`    (pure: returns owned string, no side-effects)
//!   - `emit_diff_line`   → moved to diff_utils.writeDiffLine (pure writer-based)
//!   - Global state       → eliminated; callers thread context through as parameters

const std = @import("std");

// ---------------------------------------------------------------------------
// Tool line formatting
// ---------------------------------------------------------------------------

/// Format a tool output line as `"<prefix> <text>"` or just `"<text>"` when
/// prefix is empty.
///
/// This is the pure formatting core of `tool_emit_line` from the C code,
/// without the side effects (TUI queue posting / printf).
///
/// Caller must free the result with `allocator.free`.
pub fn formatToolLine(
    allocator: std.mem.Allocator,
    prefix: []const u8,
    text: []const u8,
) ![]u8 {
    if (prefix.len > 0 and text.len > 0) {
        return std.fmt.allocPrint(allocator, "{s} {s}", .{ prefix, text });
    } else if (prefix.len > 0) {
        return allocator.dupe(u8, prefix);
    } else {
        return allocator.dupe(u8, text);
    }
}

/// Output destination for tool messages.
/// In interactive TUI mode this would route to the message queue;
/// in one-shot/subagent mode it routes to stdout or is suppressed.
pub const OutputDest = enum {
    /// Write directly to stdout (interactive non-TUI mode).
    stdout,
    /// Suppress output (oneshot/subagent mode).
    suppress,
    /// Buffer into an ArrayList for capture (testing, oneshot output).
    buffer,
};

/// Emit a tool line to the given destination.
///
/// For `OutputDest.buffer` a non-null `buf` must be provided.
pub fn emitToolLine(
    prefix: []const u8,
    text: []const u8,
    dest: OutputDest,
    buf: ?*std.ArrayList(u8),
) !void {
    switch (dest) {
        .suppress => return,
        .stdout => {
            const stdout = std.io.getStdOut().writer();
            if (prefix.len > 0 and text.len > 0) {
                try stdout.print("{s} {s}\n", .{ prefix, text });
            } else if (prefix.len > 0) {
                try stdout.print("{s}\n", .{prefix});
            } else {
                try stdout.print("{s}\n", .{text});
            }
        },
        .buffer => {
            const b = buf orelse return error.NullBuffer;
            if (prefix.len > 0 and text.len > 0) {
                try b.writer().print("{s} {s}\n", .{ prefix, text });
            } else if (prefix.len > 0) {
                try b.writer().print("{s}\n", .{prefix});
            } else {
                try b.writer().print("{s}\n", .{text});
            }
        },
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "formatToolLine: prefix + text" {
    const a = std.testing.allocator;
    const line = try formatToolLine(a, "✓", "file written");
    defer a.free(line);
    try std.testing.expectEqualStrings("✓ file written", line);
}

test "formatToolLine: empty prefix" {
    const a = std.testing.allocator;
    const line = try formatToolLine(a, "", "just text");
    defer a.free(line);
    try std.testing.expectEqualStrings("just text", line);
}

test "formatToolLine: empty text" {
    const a = std.testing.allocator;
    const line = try formatToolLine(a, "prefix", "");
    defer a.free(line);
    try std.testing.expectEqualStrings("prefix", line);
}

test "formatToolLine: both empty" {
    const a = std.testing.allocator;
    const line = try formatToolLine(a, "", "");
    defer a.free(line);
    try std.testing.expectEqualStrings("", line);
}

test "emitToolLine: buffer destination captures output" {
    var buf = std.ArrayList(u8).init(std.testing.allocator);
    defer buf.deinit();

    try emitToolLine("»", "reading file", .buffer, &buf);
    try emitToolLine("", "line two", .buffer, &buf);

    const output = buf.items;
    try std.testing.expect(std.mem.indexOf(u8, output, "» reading file") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "line two") != null);
}

test "emitToolLine: suppress produces no output" {
    // Just verifies no error is returned
    try emitToolLine("any", "thing", .suppress, null);
}
